028-network-policy-generation-feature-gate.md

```diff
@@ -0,0 +1,82 @@
+# Network Policy Generation Feature Gate
+
+The feature would consist of a new Feature Gate which can be used to disable the generation of Network Policies by Strimzi. This will be useful when a user wishes to set up a custom Network Policy set.
+
+## Current situation
+
+Currently, Strimzi generates Network Policies which allow all pods in the deployment to connect to Kafka, Zookeeper and any other service that it has generated.
+
+There is no way to stop this generation.
+
+It is possible, in the select case of listeners, to further restrict the Ingress using NetworkPolicyPeers, however this does not extend to all Network Policy generation, for example Metrics.
+It is also not possible to set up Egress Network Policies. 
+
+## Motivation
+
+The current Network Policies that are generated are great for getting Strimzi up and running in a deployment where a deny-all network policy is in place.
+
+However, this does not allow users of Strimzi to implement a custom Network Policy set, due to Strimzi not providing the full feature set of Kubernetes Network Policies.
+
+By adding a feature gate, it allows users to disable this generation, and instead write fully custom Network Policies. 
+
+When disabled the responsibility of setting up secure network policies is that of the user, and in environments where a deny-all network policy exists Strimzi will not operate until these are correctly setup.
+This will be documented as to which Network Policies exist and would have to be created.
+
+This setup is valid in 2 situations:
+
+* A deployment with no deny-all Network Policy, where all pods can already communicate with each other.
+* A deployment where the user wishes to write their own custom Network Policies.
+  * A user may have strict guidelines on the network policy requirements which differ from Strimzi's generated policies.
+
+## Proposal
+
+The proposal would be to introduce a Feature Gate which would provide the ability to turn off the generation of Network Policies by Strimzi.
+
+As the functionality already exists to generate Network Policies, a sensible naming of this gate would be `NetworkPolicies` which by default is `true`.
+If users require custom Network Policies, they can dirsable this using the regular Feature Gate syntax.
+This Feature Gate will be a long-lived Feature Gate, and is not meant to phase in a featue. Instead, it should remain so users can turn off this functionality when needed.
+
+This Feature Gate will also transition through the Alpha, Beta, GA stages. However, as this is a long-lived feature gate which the behaviour does not become default eventually the value will remain set to `true`.
```

Comment:
I wonder if we really need to go through Alpha and Beta here. Maybe we can make it GA right away? @strimzi/maintainers WDYT?

Assuming we want to take it through the phases, it would be nice to add some preliminary plan of the versions when it goes Beta and GA as we have in https://github.com/strimzi/proposals/blob/main/026-service-account-patching.md#proposal or https://github.com/strimzi/proposals/blob/main/025-control-plain-listener.md#proposal.

Reply:
From my side, I think it would be sensible to skip the Alpha, Beta stages as the behaviour does not change.
In fact, the [proposal for Feature Gates](https://github.com/strimzi/proposals/blob/main/022-feature-gates.md#:~:text=an%20example%20of%20this%20use%20case%20might%20be%20feature%20gates%20to%20disable%20automatic%20creation%20of%20pod%20disruption%20budgets%20or%20network%20policies.) uses this feature as an example of when to skip the stages.
However happy to take your decision on this.

Reply:
As discussed, we will now go with a Environment Variable approach, rather than the Feature Gate approach.